projects/optic/src/commands/capture/capture.ts

```diff
@@ -567,41 +543,78 @@ async function setup(filePath: string): Promise<string> {
   return await getCaptureStorage(resolvedPath);
 }
 
-async function startApp(
+type Bailout = { didBailout: boolean; promise: Promise<any> };
+function startApp(
   command: string,
-  dir: string,
-  readyEndpoint: string | undefined,
-  readyInterval: number,
-  readyTimeout: number,
-  targetUrl: string
-): Promise<ChildProcessWithoutNullStreams> {
+  dir: string
+): [ChildProcessWithoutNullStreams, Bailout] {
   const cmd = commandSplitter(command);
   const app = spawn(cmd.cmd, cmd.args, { detached: true, cwd: dir });
 
   app.stderr.on('data', (data) => {
     logger.error(data.toString());
   });
 
-  // TODO: this doesn't get caught by the calller but doesn't seem to leave the proxy hanging.
-  // lets find a better way.
-  app.on('exit', (code) => {
-    if (code! > 0) {
-      throw Error(`Unexpected exit from the server with exit code: ${code}`);
-    }
-  });
+  const bailout: Bailout = {
+    didBailout: false,
+    promise: new Promise((resolve, reject) => {
+      app!.on('exit', (code) => {
+        bailout.didBailout = true;
+        reject(`Server unexpectedly exited with error code ${code}`);
+      });
+    }),
+  };
+  return [app, bailout];
+}
 
+async function waitForServer(
+  bailout: Bailout,
+  readyEndpoint: string,
+  readyInterval: number,
+  readyTimeout: number,
+  targetUrl: string
+) {
   // since ready_endpoint is not required always wait one interval. without ready_endpoint,
   // ready_interval must be at least the time it takes to start the server.
   await wait(readyInterval);
 
   //
   // wait for the app to be ready
   //
-  if (readyEndpoint) {
-    const url = urljoin(targetUrl, readyEndpoint);
-    const timeout = readyTimeout || 3 * 60 * 1_000; // 3 minutes
-    await serverReady(url, readyInterval, timeout);
-  }
 
-  return app;
+  const url = urljoin(targetUrl, readyEndpoint);
+  const timeout = readyTimeout || 3 * 60 * 1_000; // 3 minutes
+  const now = Date.now();
+  const spinner = ora('Waiting for server to come online...');
+  spinner.start();
+  spinner.color = 'blue';
+
+  const checkServer = (): Promise<boolean> =>
+    fetch(url)
+      .then((res) => String(res.status).startsWith('2'))
+      .catch((e) => {
+        logger.debug(e);
+        return false;
+      });
+
+  const serverReadyPromise = new Promise(async (resolve) => {
+    let done = false;
+    // We need to bail out if the server shut down, otherwise we never conclude this promise chain
+    while (!done && !bailout.didBailout) {
```

Comment:
We need to bail out here otherwise this basically infinite loops (and node tried to wait for this to exit but it never does)